File: src/bpp_shared/world/generator/shared/cave_gen.cpp
/*
 * Copyright (c) 2025-2026, Pixel Brush <pixelbrush.dev>
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 * Based on code by Mojang Studios (2011)
*/

#include "cave_gen.h"
#include "generator/generator.h"
#include "java_math.h"

/**
 * @brief Attempts to generate a cave in the current chunk
 *
 * @param chunk The chunk to carve caves into
 * @param seed  The world seed
 */
void CaveGenerator::GenerateCavesForChunk(Chunk& chunk, int64_t seed) {
	int32_t carveExtent = this->m_carveExtentLimit;
	m_rand.setSeed(seed);
	int64_t xOffset = m_rand.nextLong() / 2L * 2L + 1L;
	int64_t zOffset = m_rand.nextLong() / 2L * 2L + 1L;

	// Use unsigned arithmetic to avoid overflow UB
	uint64_t xOffset_u = static_cast<uint64_t>(xOffset);
	uint64_t zOffset_u = static_cast<uint64_t>(zOffset);
	uint64_t seed_u = static_cast<uint64_t>(seed);

	for (int32_t cXoffset = chunk.cpos.x - carveExtent; cXoffset <= chunk.cpos.x + carveExtent; ++cXoffset) {
		for (int32_t cZoffset = chunk.cpos.z - carveExtent; cZoffset <= chunk.cpos.z + carveExtent; ++cZoffset) {
			uint64_t xPart = static_cast<uint64_t>(static_cast<int64_t>(cXoffset)) * xOffset_u;
			uint64_t zPart = static_cast<uint64_t>(static_cast<int64_t>(cZoffset)) * zOffset_u;
			uint64_t combined = (xPart + zPart) ^ seed_u;
			m_rand.setSeed(static_cast<int64_t>(combined));
			this->GenerateCaves(chunk, { cXoffset, cZoffset });
		}
	}
}

void CaveGenerator::GenerateCaves(Chunk& chunk, Int2 chunkOffset) {
	int32_t numberOfCaves = m_rand.nextInt(m_rand.nextInt(m_rand.nextInt(m_isNetherCave ? 10 : 40) + 1) + 1);
	if (m_rand.nextInt(m_isNetherCave ? 5 : 15) != 0) {
		numberOfCaves = 0;
	}

	for (int32_t caveIndex = 0; caveIndex < numberOfCaves; ++caveIndex) {
		Vec3 offset = VEC3_ZERO;
		offset.x = double(chunkOffset.x * CHUNK_WIDTH + m_rand.nextInt(CHUNK_WIDTH));
		offset.y = m_isNetherCave ? double(m_rand.nextInt(128)) : double(m_rand.nextInt(m_rand.nextInt(120) + 8));
		offset.z = double(chunkOffset.y * CHUNK_WIDTH + m_rand.nextInt(CHUNK_WIDTH));
		int32_t numberOfNodes = 1;
		if (m_rand.nextInt(4) == 0) {
			this->CarveCave(chunk, offset);
			numberOfNodes += m_rand.nextInt(4);
		}

		for (int32_t nodeIndex = 0; nodeIndex < numberOfNodes; ++nodeIndex) {
			float carveYaw = m_rand.nextFloat() * JavaMath::PI_FLOAT * 2.0f;
			float carvePitch = (m_rand.nextFloat() - 0.5f) * 2.0f / 8.0f;
			float tunnelRadius = m_rand.nextFloat() * 2.0f + m_rand.nextFloat();
			this->CarveCave(chunk, offset, m_isNetherCave ? tunnelRadius * 2.0f : tunnelRadius, carveYaw, carvePitch, 0,
			                0, 1.0);
		}
	}
}

void CaveGenerator::CarveCave(Chunk& chunk, Vec3 offset) {
	this->CarveCave(chunk, offset, 1.0f + m_rand.nextFloat() * 6.0f, 0.0f, 0.0f, -1, -1, 0.5);
}

void CaveGenerator::CarveCave(Chunk& chunk, Vec3 offset, float tunnelRadius, float carveYaw, float carvePitch,
                              int32_t tunnelStep, int32_t tunnelLength, double verticalScale) {
	double chunkCenterX = double(chunk.cpos.x * CHUNK_WIDTH + (CHUNK_WIDTH * 0.5));
	double chunkCenterZ = double(chunk.cpos.z * CHUNK_WIDTH + (CHUNK_WIDTH * 0.5));
	float pitchVel = 0.0f;
	float yawVel = 0.0f;
	Java::Random rand2(m_rand.nextLong());

	if (tunnelLength <= 0) {
		int32_t max_tunnel_length = this->m_carveExtentLimit * CHUNK_WIDTH - CHUNK_WIDTH;
		tunnelLength = max_tunnel_length - rand2.nextInt(max_tunnel_length / 4);
	}

	bool branch_tunnel = false;
	if (tunnelStep == -1) {
		tunnelStep = tunnelLength / 2;
		branch_tunnel = true;
	}

	int32_t branchPoint = rand2.nextInt(tunnelLength / 2) + tunnelLength / 4;

	bool tunnel_steepness = rand2.nextInt(6) == 0;

	if (Generator::fastGeneration) {
		double dx = offset.x - chunkCenterX;
		double dz = offset.z - chunkCenterZ;
		double remaining = double(tunnelLength - tunnelStep);
		double limit = double(tunnelRadius + 2.0f + 16.0f);
		if ((dx * dx + dz * dz - remaining * remaining) > (limit * limit))
			return;
	}

	for (; tunnelStep < tunnelLength; ++tunnelStep) {
		double radius_xz = 1.5 + double(MathHelper::sin(float(tunnelStep) * JavaMath::PI_FLOAT / float(tunnelLength)) *
		                                tunnelRadius * 1.0f);
		double radius_y = radius_xz * verticalScale;
		float pCos = MathHelper::cos(carvePitch);
		float pSin = MathHelper::sin(carvePitch);
		offset.x += double(MathHelper::cos(carveYaw) * pCos);
		offset.y += double(pSin);
		offset.z += double(MathHelper::sin(carveYaw) * pCos);

		carvePitch *= tunnel_steepness ? 0.92f : 0.7f;

		carvePitch += yawVel * 0.1f;
		carveYaw += pitchVel * 0.1f;
		yawVel *= 0.9f;
		pitchVel *= 12.0f / 16.0f;
		yawVel += (rand2.nextFloat() - rand2.nextFloat()) * rand2.nextFloat() * 2.0f;
		pitchVel += (rand2.nextFloat() - rand2.nextFloat()) * rand2.nextFloat() * 4.0f;

		if (!branch_tunnel && tunnelStep == branchPoint && tunnelRadius > 1.0f) {
			this->CarveCave(chunk, offset, rand2.nextFloat() * 0.5f + 0.5f, carveYaw - JavaMath::PI_FLOAT * 0.5f,
			                carvePitch / 3.0f, tunnelStep, tunnelLength, 1.0);
			this->CarveCave(chunk, offset, rand2.nextFloat() * 0.5f + 0.5f, carveYaw + JavaMath::PI_FLOAT * 0.5f,
			                carvePitch / 3.0f, tunnelStep, tunnelLength, 1.0);
			return;
		}

		if (branch_tunnel || rand2.nextInt(4) != 0) {
			double dx = offset.x - chunkCenterX;
			double dz = offset.z - chunkCenterZ;
			double dist = double(tunnelLength - tunnelStep);
			double limit = double(tunnelRadius + 2.0f + 16.0f);
			if ((dx * dx + dz * dz - dist * dist) > (limit * limit))
				return;

			if (offset.x >= chunkCenterX - 16.0 - radius_xz * 2.0 && offset.z >= chunkCenterZ - 16.0 - radius_xz * 2.0 &&
			    offset.x <= chunkCenterX + 16.0 + radius_xz * 2.0 && offset.z <= chunkCenterZ + 16.0 + radius_xz * 2.0) {
				int32_t xMin = MathHelper::floor_double(offset.x - radius_xz) - chunk.cpos.x * 16 - 1;
				int32_t xMax = MathHelper::floor_double(offset.x + radius_xz) - chunk.cpos.x * 16 + 1;
				int32_t yMin = MathHelper::floor_double(offset.y - radius_y) - 1;
				int32_t yMax = MathHelper::floor_double(offset.y + radius_y) + 1;
				int32_t zMin = MathHelper::floor_double(offset.z - radius_xz) - chunk.cpos.z * 16 - 1;
				int32_t zMax = MathHelper::floor_double(offset.z + radius_xz) - chunk.cpos.z * 16 + 1;

				if (xMin < 0)
					xMin = 0;
				if (xMax > 16)
					xMax = 16;
				if (yMin < 1)
					yMin = 1;
				if (yMax > 120)
					yMax = 120;
				if (zMin < 0)
					zMin = 0;
				if (zMax > 16)
					zMax = 16;

				// Check for water before carving
				bool fluidIsPresent = false;
				for (int32_t blockX = xMin; !fluidIsPresent && blockX < xMax; ++blockX) {
					for (int32_t blockZ = zMin; !fluidIsPresent && blockZ < zMax; ++blockZ) {
						for (int32_t blockY = yMax + 1; !fluidIsPresent && blockY >= yMin - 1; --blockY) {
							if (blockY >= 0 && blockY < CHUNK_HEIGHT) {
								BlockType blockType = chunk.getBlock({ blockX, blockY, blockZ });
								// Overworld caver check
								if (!m_isNetherCave &&
								    (blockType == BLOCK_WATER_FLOWING || blockType == BLOCK_WATER_STILL))
									fluidIsPresent = true;
								// Nether caver check
								if (m_isNetherCave && (blockType == BLOCK_LAVA_FLOWING || blockType == BLOCK_LAVA_STILL))
									fluidIsPresent = true;
								// Skip interior, only check the shell
								if (blockY != yMin - 1 && blockX != xMin && blockX != xMax - 1 && blockZ != zMin &&
								    blockZ != zMax - 1) {
									blockY = yMin;
								}
							}
						}
					}
				}

				if (!fluidIsPresent) {
					for (int32_t blockX = xMin; blockX < xMax; ++blockX) {
						double center_dx = (double(blockX + chunk.cpos.x * 16) + 0.5 - offset.x) / radius_xz;

						for (int32_t blockZ = zMin; blockZ < zMax; ++blockZ) {
							double center_dz = (double(blockZ + chunk.cpos.z * 16) + 0.5 - offset.z) / radius_xz;

							if (center_dx * center_dx + center_dz * center_dz < 1.0) {
								// Doesn't exist in nether caver
								bool isGrass = false;
								for (int32_t blockY = yMax - 1; blockY >= yMin; --blockY) {
									Int3 bpos{ blockX, blockY + 1, blockZ };
									double center_dy = (double(blockY) + 0.5 - offset.y) / radius_y;
									if (center_dy > -0.7 &&
									    center_dx * center_dx + center_dy * center_dy + center_dz * center_dz < 1.0) {
										BlockType blockType = chunk.getBlock(bpos);
										// Nether caver behavior
										// Dirt and grass check is most likely irrelevant,
										// but it still exists in the Vanilla Nether caver
										if (m_isNetherCave && (blockType == BLOCK_NETHERRACK ||
										                       blockType == BLOCK_DIRT || blockType == BLOCK_GRASS)) {
											chunk.setBlock(bpos, BLOCK_AIR);
											continue;
										}
										// Overworld caver behavior
										if (blockType == BLOCK_GRASS)
											isGrass = true;
										if (blockType != BLOCK_STONE && blockType != BLOCK_DIRT &&
										    blockType != BLOCK_GRASS) {
											continue;
										}
										if (blockY < 10) {
											chunk.setBlock(bpos, BLOCK_LAVA_FLOWING);
											continue;
										}
										chunk.setBlock(bpos, BLOCK_AIR);
										if (!isGrass)
											continue;
										Int3 below{ bpos.x, blockY, bpos.z };
										if (chunk.getBlock(below) == BLOCK_DIRT) {
											chunk.setBlock(below, BLOCK_GRASS);
										}
									}
								}
							}
						}
					}

					if (branch_tunnel)
						break;
				}
			}
		}
	}
}

File: crates/bpp_shared/src/world/generator/shared/cave_gen.rs
/*
 * Copyright (c) 2025-2026, Pixel Brush <pixelbrush.dev>
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 * Based on code by Mojang Studios (2011)
*/

use crate::constants::{CHUNK_HEIGHT, CHUNK_WIDTH};
use crate::enums::blocks::{
    BLOCK_AIR, BLOCK_DIRT, BLOCK_GRASS, BLOCK_LAVA_FLOWING, BLOCK_LAVA_STILL, BLOCK_NETHERRACK, BLOCK_STONE,
    BLOCK_WATER_FLOWING, BLOCK_WATER_STILL,
};
use crate::helpers::java::java_math::{JavaMath, MathHelper};
use crate::helpers::java::java_random::Random;
use crate::numeric_structs::{Int2, Int3, VEC3_ZERO, Vec3};
use crate::world::chunk::Chunk;

/**
 * @brief Used to carve caves into the world
 *
 */
pub struct CaveGenerator {
    rand: Random,
    pub is_nether_cave: bool,
}

impl CaveGenerator {
    const CARVE_EXTENT_LIMIT: i32 = 8;

    pub fn new(is_nether_cave: bool) -> Self {
        Self { rand: Random::new(), is_nether_cave }
    }

    /**
     * @brief Attempts to generate a cave in the current chunk
     *
     * @param chunk The chunk to carve caves into
     * @param seed  The world seed
     */
    pub fn generate_caves_for_chunk(&mut self, chunk: &mut Chunk, seed: i64) {
        let carve_extent = Self::CARVE_EXTENT_LIMIT;
        self.rand.set_seed(seed);
        let x_offset = self.rand.next_long() / 2i64 * 2i64 + 1i64;
        let z_offset = self.rand.next_long() / 2i64 * 2i64 + 1i64;

        // Use unsigned arithmetic to avoid overflow UB
        let x_offset_u = x_offset as u64;
        let z_offset_u = z_offset as u64;
        let seed_u = seed as u64;

        for c_x_offset in (chunk.cpos.x - carve_extent)..=(chunk.cpos.x + carve_extent) {
            for c_z_offset in (*chunk.cpos.z() - carve_extent)..=(*chunk.cpos.z() + carve_extent) {
                let x_part = (c_x_offset as i64 as u64).wrapping_mul(x_offset_u);
                let z_part = (c_z_offset as i64 as u64).wrapping_mul(z_offset_u);
                let combined = x_part.wrapping_add(z_part) ^ seed_u;
                self.rand.set_seed(combined as i64);
                self.generate_caves(chunk, Int2::new(c_x_offset, c_z_offset));
            }
        }
    }

    pub fn generate_caves(&mut self, chunk: &mut Chunk, chunk_offset: Int2) {
        let cave_bound_a = self.rand.next_int_bound(if self.is_nether_cave { 10 } else { 40 });
        let cave_bound_b = self.rand.next_int_bound(cave_bound_a + 1);
        let mut number_of_caves = self.rand.next_int_bound(cave_bound_b + 1);
        if self.rand.next_int_bound(if self.is_nether_cave { 5 } else { 15 }) != 0 {
            number_of_caves = 0;
        }

        for _cave_index in 0..number_of_caves {
            let mut offset = VEC3_ZERO;
            offset.x = f64::from(chunk_offset.x * CHUNK_WIDTH + self.rand.next_int_bound(CHUNK_WIDTH));
            offset.y = if self.is_nether_cave {
                f64::from(self.rand.next_int_bound(128))
            } else {
                let inner = self.rand.next_int_bound(120);
                f64::from(self.rand.next_int_bound(inner + 8))
            };
            offset.z = f64::from(chunk_offset.y * CHUNK_WIDTH + self.rand.next_int_bound(CHUNK_WIDTH));
            let mut number_of_nodes = 1;
            if self.rand.next_int_bound(4) == 0 {
                self.carve_cave(chunk, offset);
                number_of_nodes += self.rand.next_int_bound(4);
            }

            for _node_index in 0..number_of_nodes {
                let carve_yaw = self.rand.next_float() * JavaMath::PI_FLOAT * 2.0;
                let carve_pitch = (self.rand.next_float() - 0.5) * 2.0 / 8.0;
                let tunnel_radius = self.rand.next_float() * 2.0 + self.rand.next_float();
                self.carve_cave_full(
                    chunk,
                    offset,
                    if self.is_nether_cave { tunnel_radius * 2.0 } else { tunnel_radius },
                    carve_yaw,
                    carve_pitch,
                    0,
                    0,
                    1.0,
                );
            }
        }
    }

    pub fn carve_cave(&mut self, chunk: &mut Chunk, offset: Vec3) {
        let tunnel_radius = 1.0 + self.rand.next_float() * 6.0;
        self.carve_cave_full(chunk, offset, tunnel_radius, 0.0, 0.0, -1, -1, 0.5);
    }

    pub fn carve_cave_full(
        &mut self,
        chunk: &mut Chunk,
        mut offset: Vec3,
        tunnel_radius: f32,
        mut carve_yaw: f32,
        mut carve_pitch: f32,
        mut tunnel_step: i32,
        mut tunnel_length: i32,
        vertical_scale: f64,
    ) {
        let chunk_center_x = f64::from(chunk.cpos.x * CHUNK_WIDTH) + CHUNK_WIDTH as f64 * 0.5;
        let chunk_center_z = f64::from(*chunk.cpos.z() * CHUNK_WIDTH) + CHUNK_WIDTH as f64 * 0.5;
        let mut pitch_vel = 0.0f32;
        let mut yaw_vel = 0.0f32;
        let mut rand2 = Random::with_seed(self.rand.next_long());

        if tunnel_length <= 0 {
            let max_tunnel_length = Self::CARVE_EXTENT_LIMIT * CHUNK_WIDTH - CHUNK_WIDTH;
            tunnel_length = max_tunnel_length - rand2.next_int_bound(max_tunnel_length / 4);
        }

        let mut branch_tunnel = false;
        if tunnel_step == -1 {
            tunnel_step = tunnel_length / 2;
            branch_tunnel = true;
        }

        let branch_point = rand2.next_int_bound(tunnel_length / 2) + tunnel_length / 4;

        let tunnel_steepness = rand2.next_int_bound(6) == 0;
        while tunnel_step < tunnel_length {
            let radius_xz = 1.5
                + f64::from(
                    MathHelper::sin(tunnel_step as f32 * JavaMath::PI_FLOAT / tunnel_length as f32) * tunnel_radius * 1.0,
                );
            let radius_y = radius_xz * vertical_scale;
            let p_cos = MathHelper::cos(carve_pitch);
            let p_sin = MathHelper::sin(carve_pitch);
            offset.x += f64::from(MathHelper::cos(carve_yaw) * p_cos);
            offset.y += f64::from(p_sin);
            offset.z += f64::from(MathHelper::sin(carve_yaw) * p_cos);

            carve_pitch *= if tunnel_steepness { 0.92 } else { 0.7 };

            carve_pitch += yaw_vel * 0.1;
            carve_yaw += pitch_vel * 0.1;
            yaw_vel *= 0.9;
            pitch_vel *= 12.0 / 16.0;
            yaw_vel += (rand2.next_float() - rand2.next_float()) * rand2.next_float() * 2.0;
            pitch_vel += (rand2.next_float() - rand2.next_float()) * rand2.next_float() * 4.0;

            if !branch_tunnel && tunnel_step == branch_point && tunnel_radius > 1.0 {
                self.carve_cave_full(
                    chunk,
                    offset,
                    rand2.next_float() * 0.5 + 0.5,
                    carve_yaw - JavaMath::PI_FLOAT * 0.5,
                    carve_pitch / 3.0,
                    tunnel_step,
                    tunnel_length,
                    1.0,
                );
                self.carve_cave_full(
                    chunk,
                    offset,
                    rand2.next_float() * 0.5 + 0.5,
                    carve_yaw + JavaMath::PI_FLOAT * 0.5,
                    carve_pitch / 3.0,
                    tunnel_step,
                    tunnel_length,
                    1.0,
                );
                return;
            }

            if branch_tunnel || rand2.next_int_bound(4) != 0 {
                let dx = offset.x - chunk_center_x;
                let dz = offset.z - chunk_center_z;
                let dist = f64::from(tunnel_length - tunnel_step);
                let limit = f64::from(tunnel_radius + 2.0 + 16.0);
                if (dx * dx + dz * dz - dist * dist) > (limit * limit) {
                    return;
                }

                if offset.x >= chunk_center_x - 16.0 - radius_xz * 2.0
                    && offset.z >= chunk_center_z - 16.0 - radius_xz * 2.0
                    && offset.x <= chunk_center_x + 16.0 + radius_xz * 2.0
                    && offset.z <= chunk_center_z + 16.0 + radius_xz * 2.0
                {
                    let mut x_min = MathHelper::floor_double(offset.x - radius_xz) - chunk.cpos.x * 16 - 1;
                    let mut x_max = MathHelper::floor_double(offset.x + radius_xz) - chunk.cpos.x * 16 + 1;
                    let mut y_min = MathHelper::floor_double(offset.y - radius_y) - 1;
                    let mut y_max = MathHelper::floor_double(offset.y + radius_y) + 1;
                    let mut z_min = MathHelper::floor_double(offset.z - radius_xz) - *chunk.cpos.z() * 16 - 1;
                    let mut z_max = MathHelper::floor_double(offset.z + radius_xz) - *chunk.cpos.z() * 16 + 1;

                    if x_min < 0 {
                        x_min = 0;
                    }
                    if x_max > 16 {
                        x_max = 16;
                    }
                    if y_min < 1 {
                        y_min = 1;
                    }
                    if y_max > 120 {
                        y_max = 120;
                    }
                    if z_min < 0 {
                        z_min = 0;
                    }
                    if z_max > 16 {
                        z_max = 16;
                    }

                    // Check for water before carving
                    let mut fluid_is_present = false;
                    let mut block_x = x_min;
                    while !fluid_is_present && block_x < x_max {
                        let mut block_z = z_min;
                        while !fluid_is_present && block_z < z_max {
                            let mut block_y = y_max + 1;
                            while !fluid_is_present && block_y >= y_min - 1 {
                                if block_y >= 0 && block_y < CHUNK_HEIGHT {
                                    let block_type = chunk.get_block(Int3::new(block_x, block_y, block_z));
                                    // Overworld caver check
                                    if !self.is_nether_cave
                                        && (block_type == BLOCK_WATER_FLOWING || block_type == BLOCK_WATER_STILL)
                                    {
                                        fluid_is_present = true;
                                    }
                                    // Nether caver check
                                    if self.is_nether_cave
                                        && (block_type == BLOCK_LAVA_FLOWING || block_type == BLOCK_LAVA_STILL)
                                    {
                                        fluid_is_present = true;
                                    }
                                    // Skip interior, only check the shell
                                    if block_y != y_min - 1
                                        && block_x != x_min
                                        && block_x != x_max - 1
                                        && block_z != z_min
                                        && block_z != z_max - 1
                                    {
                                        block_y = y_min;
                                    }
                                }
                                block_y -= 1;
                            }
                            block_z += 1;
                        }
                        block_x += 1;
                    }

                    if !fluid_is_present {
                        for block_x in x_min..x_max {
                            let center_dx = (f64::from(block_x + chunk.cpos.x * 16) + 0.5 - offset.x) / radius_xz;

                            for block_z in z_min..z_max {
                                let center_dz = (f64::from(block_z + *chunk.cpos.z() * 16) + 0.5 - offset.z) / radius_xz;

                                if center_dx * center_dx + center_dz * center_dz < 1.0 {
                                    // Doesn't exist in nether caver
                                    let mut is_grass = false;
                                    for block_y in (y_min..y_max).rev() {
                                        let bpos = Int3::new(block_x, block_y + 1, block_z);
                                        let center_dy = (f64::from(block_y) + 0.5 - offset.y) / radius_y;
                                        if center_dy > -0.7
                                            && center_dx * center_dx + center_dy * center_dy + center_dz * center_dz < 1.0
                                        {
                                            let block_type = chunk.get_block(bpos);
                                            // Nether caver behavior
                                            // Dirt and grass check is most likely irrelevant,
                                            // but it still exists in the Vanilla Nether caver
                                            if self.is_nether_cave
                                                && (block_type == BLOCK_NETHERRACK
                                                    || block_type == BLOCK_DIRT
                                                    || block_type == BLOCK_GRASS)
                                            {
                                                chunk.set_block(bpos, BLOCK_AIR);
                                                continue;
                                            }
                                            // Overworld caver behavior
                                            if block_type == BLOCK_GRASS {
                                                is_grass = true;
                                            }
                                            if block_type != BLOCK_STONE && block_type != BLOCK_DIRT && block_type != BLOCK_GRASS
                                            {
                                                continue;
                                            }
                                            if block_y < 10 {
                                                chunk.set_block(bpos, BLOCK_LAVA_FLOWING);
                                                continue;
                                            }
                                            chunk.set_block(bpos, BLOCK_AIR);
                                            if !is_grass {
                                                continue;
                                            }
                                            let below = Int3::new(bpos.x, block_y, bpos.z);
                                            if chunk.get_block(below) == BLOCK_DIRT {
                                                chunk.set_block(below, BLOCK_GRASS);
                                            }
                                        }
                                    }
                                }
                            }
                        }

                        if branch_tunnel {
                            break;
                        }
                    }
                }
            }

            tunnel_step += 1;
        }
    }
}
